pkg/raft/raft.go

```diff
@@ -2324,3 +2337,12 @@ func (r *raft) testingStepDown() error {
 	r.becomeFollower(r.Term, r.id) // mirror the logic in how we step down when CheckQuorum fails
 	return nil
 }
+
+// onlyAdvanceCommitViaMsgApp returns true if the commit index is advanced on
+// the followers using MsgApp only. This means that heartbeats are not used to
+// advance the commit index. This function returns true only if followers
+// understand that heartbeats might not contain a commit index.
+func (r *raft) advanceCommitViaMsgAppOnly() bool {
```

Comment:
nit: I think the precondition (that we're gating against) is not `if followers understand that heartbeats might not contain a commit index`, but rather: `if all followers communicate their durable commit index back to the leader via MsgAppResp`.

How about something like:
```
advanceCommitViaMsgAppOnly returns true if the leader can stop communicating
the commit index to followers using MsgHeartbeat, and instead do it reliably
via MsgApp. This is safe when all followers send their durable commit index
back to the leader via MsgAppResp.
```